### PR TITLE
Use AsyncStream in DasCustody

### PR DESCRIPTION
**Pull request by Nashatyrev**
## PR Description

- Implemented `AsyncStream` 
- Use stream approach for `DataColumnSidecarCustody.retrieveMissingColumns()`

### CODE REVIEW

File: infrastructure/async/src/main/java/tech/pegasys/teku/infrastructure/async/stream/AsyncIteratorCallback.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+package tech.pegasys.teku.infrastructure.async.stream;
+
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+
+interface AsyncIteratorCallback<T> {
+
+  SafeFuture<Boolean> TRUE_FUTURE = SafeFuture.completedFuture(true);
+  SafeFuture<Boolean> FALSE_FUTURE = SafeFuture.completedFuture(false);
+
+  SafeFuture<Boolean> onNext(T t);
```
Comment (zilm13) [inline infrastructure/async/src/main/java/tech/pegasys/teku/infrastructure/async/stream/AsyncIteratorCallback.java:33]: Could you add here what's returned, please

File: infrastructure/async/src/test/java/tech/pegasys/teku/infrastructure/async/stream/AsyncIteratorTest.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    futures.get(1).complete(1);
+
+    assertThat(collector).isEmpty();
+
+    futures.get(0).complete(0);
+
+    assertThat(collector).containsExactly(0, 20, 40, 100, 120, 140);
+    assertThat(listPromise).isNotDone();
+
+    futures.get(4).completeExceptionally(new RuntimeException("test"));
```
Comment (zilm13) [inline infrastructure/async/src/test/java/tech/pegasys/teku/infrastructure/async/stream/AsyncIteratorTest.java]: I'd add clarification comment, that exception should not break all stream

File: infrastructure/async/src/test/java/tech/pegasys/teku/infrastructure/async/stream/AsyncIteratorTest.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    assertThat(listPromise).isNotDone();
+
+    futures.get(2).complete(2);
+
+    assertThat(collector).containsExactly(0, 20, 40, 100, 120, 140, 200, 220, 240);
+    assertThat(listPromise).isNotDone();
+
+    futures.get(3).complete(3);
+
+    assertThat(collector).containsExactly(0, 20, 40, 100, 120, 140, 200, 220, 240, 300);
```
Comment (zilm13) [inline infrastructure/async/src/test/java/tech/pegasys/teku/infrastructure/async/stream/AsyncIteratorTest.java]: I'd add clarification comment, that limit kicks in here

File: ethereum/statetransition/src/testFixtures/java/tech/pegasys/teku/statetransition/datacolumns/DasCustodyStand.java [hunk 4]
```diff
@@ -120,6 +120,12 @@ public Collection<UInt64> getCustodyColumnIndexes(UInt64 slot) {
         .orElse(Collections.emptyList());
   }
 
+  public int getCustodyPeriodSlots() {
+    int custodyPeriodEpochs = config.getCustodyRequirement();
```
Comment (zilm13) [inline ethereum/statetransition/src/testFixtures/java/tech/pegasys/teku/statetransition/datacolumns/DasCustodyStand.java]: wanted `getMinEpochsForDataColumnSidecarsRequests` here?

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarCustodyImpl.java [hunk 5]
```diff
@@ -64,6 +66,10 @@ public Collection<DataColumnIdentifier> getIncompleteColumns() {
           .toList();
     }
 
+    public AsyncStream<DataColumnIdentifier> steamIncompleteColumns() {
```
Comment (zilm13) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarCustodyImpl.java]: stream

**Review by zilm13 — APPROVED** [review]
Brilliant, just few nits

File: infrastructure/async/src/main/java/tech/pegasys/teku/infrastructure/async/stream/AsyncIteratorCallback.java (hunk 1 above)
Comment (Nashatyrev) [reply infrastructure/async/src/main/java/tech/pegasys/teku/infrastructure/async/stream/AsyncIteratorCallback.java:33]: Yeah, added some javadocs there

File: ethereum/statetransition/src/testFixtures/java/tech/pegasys/teku/statetransition/datacolumns/DasCustodyStand.java (hunk 4 above)
Comment (Nashatyrev) [reply ethereum/statetransition/src/testFixtures/java/tech/pegasys/teku/statetransition/datacolumns/DasCustodyStand.java]: Removed this method, it's obsolete

**Pull request merged**